provider/sns/broker_test.go

```diff
@@ -0,0 +1,210 @@
+			topics, tErr := broker.Topics(ctx)
+
+			require.NoError(t, tErr)
+			require.Len(t, topics, 1)
+			require.EqualValues(t, topicARN, topics[0].String())
+		})
+
+		consumer1 := &consumer{}
+		consumer2 := &consumer{}
+
+		t.Run("WHEN registering two subscribers to such topic", func(t *testing.T) {
+			sub1 := pubsub.NewSubscriber(consumer1.handle)
+			require.NoError(t, broker.Subscribe(ctx, topic, sub1))
+
+			sub2 := pubsub.NewSubscriber(consumer2.handle)
+			require.NoError(t, broker.Subscribe(ctx, topic, sub2))
+
+			t.Run("THEN sns registers only one subscription as they are the same topic", func(t *testing.T) {
+				subs, lErr := snsCli.ListSubscriptions(ctx, &awssns.ListSubscriptionsInput{})
+
+				require.NoError(t, lErr)
+				require.Len(t, subs.Subscriptions, 1)
+				require.EqualValues(t, topicARN, *subs.Subscriptions[0].TopicArn)
+			})
+		})
+
+		t.Run("WHEN publishing N messages to the topic", func(t *testing.T) {
+			sent := []string{
+				"hello world",
+				"lorem ipsum",
+				"dolorem sit amet",
+			}
+
+			for _, msg := range sent {
+				require.NoError(t, broker.Publish(ctx, topic, msg))
+			}
+
+			t.Run("THEN subscribers of such topic eventually receives the messages only once", func(t *testing.T) {
+				require.Eventually(t, func() bool { return consumer1.hasExactlyOnce(sent) }, 10*time.Second, time.Millisecond*100)
+				require.Eventually(t, func() bool { return consumer2.hasExactlyOnce(sent) }, 10*time.Second, time.Millisecond*100)
+			})
+		})
+	})
+}
+
+func TestMultiBroker(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	t.Run("GIVEN a sns topic and two brokers (B1 and B2) with one subscriber each", func(t *testing.T) {
+		cfg := awsConfig(t)
+		snsCli := awssns.NewFromConfig(cfg)
+		sqsCli := awssqs.NewFromConfig(cfg)
+
+		broker1, err := prepareBroker(ctx, sqsCli, snsCli, "test-queue-b1")
+		require.NoError(t, err)
+		require.NotNil(t, broker1)
+
+		broker2, err := prepareBroker(ctx, sqsCli, snsCli, "test-queue-b2")
+		require.NoError(t, err)
+		require.NotNil(t, broker2)
+
+		tRes, err := snsCli.CreateTopic(ctx, &awssns.CreateTopicInput{Name: aws.String("test-topic")})
+		require.NoError(t, err)
+		topicARN := *tRes.TopicArn
+		topic := pubsub.Topic(topicARN)
+
+		consumer1 := &consumer{}
+		sub1 := pubsub.NewSubscriber(consumer1.handle)
+		require.NoError(t, broker1.Subscribe(ctx, topic, sub1))
+
+		consumer2 := &consumer{}
+		sub2 := pubsub.NewSubscriber(consumer2.handle)
+		require.NoError(t, broker2.Subscribe(ctx, topic, sub2))
+
+		t.Run("WHEN publishing N messages to the topic using B1", func(t *testing.T) {
+			sent := []string{
+				"hello world",
+				"lorem ipsum",
+				"dolorem sit amet",
+			}
+
+			for _, msg := range sent {
+				require.NoError(t, broker1.Publish(ctx, topic, msg))
+			}
+
+			t.Run("THEN subscribers of such topic eventually receives the messages only once", func(t *testing.T) {
```

Comment:
as far as i know SQS guarantees that a message will be delivered but you can receive the same message more that once. So the assertion is true in tests but in a real environment can be violated.

Reply:
totally right. There are multiple escenarios on which this may happen, not only caused by SQS. For example, you successfully handle a message locally, but fail to notify to SQS to remove it (e.g. network error), you will receive the same message on the next poll.

In other hand, SQS can be configured for [excatly once] delivery(https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues-exactly-once-processing.html) with 5min deduplication window.

Reply:
on standard queues (order is not guaranteed, at-least-one delivery policy) this feature can not be configured, and 1 single message can be received more than once. Maybe will be a good point to store in memory, the message's ID for a short period of time (seconds) and offer a library that add's this feature to amazon standard queues, this feature avoids the network partition you mention in your previous comment. But yes, this extra bonus point would be out of focus of this library

Reply:
there no need, this is handled automatically by SNS Topic configuration:
![image](https://user-images.githubusercontent.com/749463/141486741-530c03ee-ca92-419d-b4bd-4e547025776e.png)